File: measure-android/measure/src/main/java/sh/measure/sample/exceptions/ExceptionData.kt
package sh.measure.sample.exceptions

import kotlinx.serialization.Serializable

/**
 * Factory for creating [ExceptionData].
 */
internal object ExceptionFactory {
    /**
     * Creates [ExceptionData] from a [Throwable].
     *
     * @param throwable The [Throwable] to create the [ExceptionData] from.
     * @param handled Whether the exception was handled or not.
     */
    fun createExceptionData(throwable: Throwable, handled: Boolean): ExceptionData {
        val exceptions = mutableListOf<MeasureException>()
        var error: Throwable? = throwable
        while (error != null) {
            val exception = MeasureException(
                type = error.javaClass.name,
                message = error.message,
                stackframes = error.stackTrace.map {
                    Stackframe(
                        class_name = it.className,
                        method_name = it.methodName,
                        file_name = it.fileName,
                        line_number = it.lineNumber,
                    )
                },
            )
            exceptions.add(exception)
            error = error.cause
        }
        return ExceptionData(exceptions, handled)
    }
}

/**
 * Represents an exception in Measure. This is used to track handled and unhandled exceptions.
 */
@Serializable
internal data class ExceptionData(
    /**
     * A list of exceptions that were thrown. Multiple exceptions represent "chained" exceptions.
     */
    val exceptions: List<MeasureException>,

    /**
     * Whether the exception was handled or not.
     */
    val handled: Boolean
)

/**
 * Represents a stacktrace in Measure.
 */
@Serializable
internal data class MeasureException(
    /**
     * The fully qualified type of the exception. For example, java.lang.Exception.
     */
    val type: String,

    /**
     * A message which describes the exception.
     */
    val message: String? = null,

    /**
     * A list of stack frames for the exception.
     */
    val stackframes: List<Stackframe>
)

/**
 * Represents a stackframe in Measure.
 */
@Serializable
internal data class Stackframe(
    /**
     * The fully qualified class name.
     */
    val class_name: String? = null,

    /**
     * The name of the method in the stacktrace.
     */
    val method_name: String? = null,

    /**
     * The name of the source file in the stacktrace.
     */
    val file_name: String? = null,

    /**
     * The line number of the method called.
     */
    val line_number: Int? = null
)

File: flutter/measure_flutter/lib/src/exception/exception_data.dart
import '../method_channel/method_constants.dart';

/// Represents an exception in Measure. This is used to track handled and unhandled exceptions.
class ExceptionData {
  /// A list of exceptions that were thrown. Multiple exceptions represent "chained" exceptions.
  final List<ExceptionUnit> exceptions;

  /// Whether the exception was handled or not.
  final bool handled;

  ExceptionData({
    required this.exceptions,
    required this.handled,
  });

  Map<String, dynamic> toJson() {
    return _serializeExceptionData(this);
  }

  Map<String, dynamic> _serializeExceptionData(ExceptionData data) {
    return {
      MethodConstants.exceptionExceptions:
      data.exceptions.map((e) => _serializeExceptionUnit(e)).toList(),
      MethodConstants.exceptionHandled: data.handled,
    };
  }

  Map<String, dynamic> _serializeExceptionUnit(ExceptionUnit unit) {
    return {
      MethodConstants.exceptionType: unit.type,
      MethodConstants.exceptionMessage: unit.message,
      MethodConstants.exceptionFrames:
      unit.frames.map((f) => _serializeFrame(f)).toList(),
    };
  }

  Map<String, dynamic> _serializeFrame(MsrFrame frame) {
    return {
      MethodConstants.exceptionFrameClassName: frame.className,
      MethodConstants.exceptionFrameMethodName: frame.methodName,
      MethodConstants.exceptionFrameFileName: frame.fileName,
      MethodConstants.exceptionFrameLineNum: frame.lineNum,
      MethodConstants.exceptionFrameModuleName: frame.moduleName,
      MethodConstants.exceptionFrameColNum: frame.colNum,
      MethodConstants.exceptionFrameIndex: frame.frameIndex,
      MethodConstants.exceptionFrameBinaryAddr: frame.binaryAddr,
      MethodConstants.exceptionFrameInstructionAddr: frame.instructionAddr,
    };
  }
}

/// Represents a stacktrace in Measure.
class ExceptionUnit {
  /// The fully qualified type of the exception. For example, java.lang.Exception.
  final String? type;

  /// A message which describes the exception.
  final String? message;

  /// A list of stack frames for the exception.
  final List<MsrFrame> frames;

  ExceptionUnit({
    this.type,
    this.message,
    required this.frames,
  });
}

/// Represents a stack frame in Measure.
class MsrFrame {
  /// The fully qualified class name.
  final String? className;

  /// The name of the method in the stacktrace.
  final String? methodName;

  /// The name of the source file in the stacktrace.
  final String? fileName;

  /// The line number of the method called.
  final int? lineNum;

  final String? moduleName;

  final int? colNum;

  final int? frameIndex;

  final String? binaryAddr;

  final String? symbolAddress;

  final String? instructionAddr;

  MsrFrame({
    this.className,
    this.methodName,
    this.fileName,
    this.lineNum,
    this.moduleName,
    this.colNum,
    this.frameIndex,
    this.binaryAddr,
    this.symbolAddress,
    this.instructionAddr,
  });
}